File: Project1.py


StringToInt = {
    "0": 0,
    "1": 1,
    "2": 2,
    "3": 3,
    "4": 4,
    "5": 5,
    "6": 6,
    "7": 7,
    "8": 8,
    "9": 9,
}

def isNumber(char)->bool:
    if char == "0" or char == "1" or char == "2" or char == "3" or char == "4" or char == "5" or char == "6" or char == "7" or char == "8" or char == "9" or char == "_": 
        return True
    else:
        return False
    
def isFloat(number)->bool:

    if number[len(number)-1] == "f" or number[len(number)-1] == "F":
        return True
    else:
        return False
    
def isWithinFloatRange(number)->bool:
    return True

def stringToNumber(stringNumber):
    finalNumber = 0

    index = len(wholeNumber)-1

    multiplier = 1

    while index >= 0:
        finalNumber += StringToInt[wholeNumber[index]]*multiplier
        multiplier *= 10
        index -= 1
        
    return finalNumber*1.0
    

def addWholeAndDecimal(wholeNumber, decimalNumber):

    finalNumber = 0

    index = len(wholeNumber)-1

    multiplier = 1

    while index >= 0:
        finalNumber += StringToInt[wholeNumber[index]]*multiplier
        multiplier *= 10
        index -= 1

    if decimalNumber != "":
        index = 0
        multiplier = 0.1
        while index < len(decimalNumber):
            finalNumber += StringToInt[decimalNumber[index]]*multiplier
            multiplier /= 10
            index += 1

    return finalNumber*1.0
        
def exponentValue(floatValue, exponentNumber):

    initialFloatValue = floatValue

    index = len(exponentNumber)-1
    multiplier = 1
    exponentValue = 0

    while index >= 0:
        exponentValue += StringToInt[exponentNumber[index]]*multiplier
        multiplier *= 10
        index -= 1


    if exponentValue == 0:
        return floatValue*0.1

    while exponentValue > 0:
        floatValue *= 10
        exponentValue -= 1


    return floatValue*0.1

def main(number)->float:

    wholeNumber = ""
    decimalNumber = ""
    exponentNumber = ""

    if isFloat(number):
        
        if isWithinFloatRange(number):

            ## Retrieve Whole Number (and take out all the underscores)

            index = 0
            loop = True
            wholeNumberOnly = False
            
            while loop is True and index < len(number):

                if number[index] == ".":
                    loop = False
                elif number[index] == "e" or number[index] == "E":
                    wholeNumberOnly = True
                    loop = False
                elif number[index] == "_":
                    if index == 0 or index == len(number)-1 or not isNumber(number[index-1]) or not isNumber(number[index+1]):
                        return -1
                elif index == len(number)-1 and (number[index] == "f" or number[index] != "F"):
                    loop = False
                    return stringToNumber(wholeNumber)
                elif not isNumber(number[index]):
                    return -1   
                else:
                    wholeNumber += number[index]
                    
                index+= 1

            ## Retrieve decimal number (and take out all underscores)

            loop = True

            if wholeNumberOnly is False:
                while loop is True and index < len(number):
                    
                    if number[index] == ".":
                        return -1
                    elif number[index] == "e" or number[index] == "E":
                        loop = False
                    elif number[index] == "_":
                        if index == len(number)-1 or not isNumber(number[index-1]) or not isNumber(number[index+1]):
                            return -1
                    elif index == len(number)-1 and (number[index] == "f" or number[index] == "F"):
                        return addWholeAndDecimal(wholeNumber, decimalNumber)
                    elif not isNumber(number[index]):
                        return -1
                    else:
                        decimalNumber += number[index]
                    index+=1

            floatValue = addWholeAndDecimal(wholeNumber, decimalNumber)

            ## Retrieve exponent number

            loop = True

            while loop is True and index < len(number):
                if number[index] == ".":
                    return -1
                elif number[index] == "e" or number[index] == "E":
                    return -1
                elif number[index] == "_":
                    if index == len(number)-1 or not isNumber(number[index-1]) or not isNumber(number[index+1]):
                        return -1
                elif index == len(number)-1 and (number[index] == "f" or number[index] == "F"):
                    if exponentNumber == "":
                        return -1
                    else:
                        return exponentValue(floatValue, exponentNumber)
                    
                elif not isNumber(number[index]):
                    return -1
                else:
                    exponentNumber += number[index]
                index+=1

    return -1

if __name__ == "__main__": 
    while True:
        print("Enter number: ", end='')
        number = input()
        if number != "-1":
            result = main(number)
            if result == -1:
                print("Invalid Result")
            else:
                print(result)
            print()
        else:
            break

File: Project2/src/com/company/Project1.java
package com.company;

import static com.company.Calculator.isNumber;

public class Project1 {
    public float convertStringToFloat(String tempFloatValue) {
        String wholeNumber = "";
        String decimalNumber = "";
        String exponentNumber = "";
        float floatValue = 0.0f;

        if(isFloat(tempFloatValue))
        {
            int index = 0;
            boolean loop = true;
            boolean wholeNumberOnly = false;
            boolean exponent = false;

            while(loop && index < tempFloatValue.length()) {
                switch (tempFloatValue.charAt(index)) {
                    case '.':
                        loop = false;
                        break;
                    case 'e', 'E':
                        wholeNumberOnly = true;
                        exponent = true;
                        loop = false;
                        break;
                    case '_':
                        if (index == 0 ||
                                index == tempFloatValue.length() - 1 ||
                                !isNumber(tempFloatValue.charAt(index - 1)) ||
                                !isNumber(tempFloatValue.charAt(index + 1)))
                            return -1f;
                        break;
                    case 'f', 'F':
                        if (index == tempFloatValue.length() - 1) {
                            loop = false;
                            break;
                        } else
                            return -1f;
                    default:
                        if (!isNumber(tempFloatValue.charAt(index)))
                            return -1f;
                        else
                            wholeNumber += tempFloatValue.charAt(index);
                        break;
                }
                index += 1;
            }
            loop = true;

            if(wholeNumberOnly == false)
            {
                while(loop && index < tempFloatValue.length())
                {
                    switch(tempFloatValue.charAt(index))
                    {
                        case '.':
                            return -1f;
                        case 'e', 'E':
                            loop = false;
                            exponent = true;
                            break;
                        case '_':
                            if(index == tempFloatValue.length()-1 ||
                                    !isNumber(tempFloatValue.charAt(index-1)) ||
                                    !isNumber(tempFloatValue.charAt(index+1)))
                                return -1f;
                            break;
                        case 'f', 'F':
                            if(index == tempFloatValue.length()-1)
                            {
                                return stringToNumber(wholeNumber, decimalNumber);
                            }
                            else
                                return -1;
                        default:
                            if(!isNumber(tempFloatValue.charAt(index)))
                                return -1;
                            else
                                decimalNumber += tempFloatValue.charAt(index);
                            break;
                    }
                    index += 1;
                }
                floatValue = stringToNumber(wholeNumber, decimalNumber);
            }
            else
                floatValue += stringToNumber(wholeNumber, null);

            if(exponent == true)
            {
                loop = true;
                char exponentSign = '+';

                if(index < tempFloatValue.length())
                {
                    if(tempFloatValue.charAt(index) == '-')
                    {
                        exponentSign = '-';
                        index++;
                    }
                    else if (tempFloatValue.charAt(index) == '+')
                        index++;
                }

                while(loop && index < tempFloatValue.length())
                {
                    switch(tempFloatValue.charAt(index))
                    {
                        case '-', '+', '.', 'e', 'E':
                            return -1;
                        case '_':
                            if(index == tempFloatValue.length()-1 ||
                                    !isNumber(tempFloatValue.charAt(index-1)) ||
                                    !isNumber(tempFloatValue.charAt(index+1)))
                                return -1;
                        case 'f', 'F':
                            if(index == tempFloatValue.length()-1)
                                if(exponentNumber=="")
                                    return -1;
                            break;
                        default:
                            if(!isNumber(tempFloatValue.charAt(index)))
                                return -1;
                            else
                                exponentNumber += tempFloatValue.charAt(index);

                    }
                    index++;
                }

                floatValue = exponentValue(floatValue, exponentNumber, exponentSign);
            }
            return floatValue;
        }
        return -1f;
    }

    private float exponentValue(float floatValue, String exponentNumber, char exponentSign) {
        int index = exponentNumber.length()-1;
        int multiplier = 1;
        float exponentValue = stringToNumber(exponentNumber, null);

        if(exponentValue == 0.0f)
            return floatValue/10;
        else if( exponentSign == '+')
            while(exponentValue > 0)
            {
                floatValue*= 10;
                exponentValue--;
            }
        else if(exponentSign == '-')
            while(exponentValue > 0)
            {
                floatValue/=10;
                exponentValue--;
            }
        else
            return -1;

        return floatValue;
    }

    private float stringToNumber(String wholeNumber, String decimalNumber) {
        float finalNumber = 0.0f;
        int index = wholeNumber.length()-1;
        int multiplier = 1;

        while(index >= 0)
        {
            finalNumber += CharToFloat(wholeNumber.charAt(index))*multiplier;
            multiplier *= 10;
            index--;
        }

        if(decimalNumber != null)
        {
            index = 0;
            multiplier = 10;

            while(index < decimalNumber.length())
            {
                finalNumber += CharToFloat(decimalNumber.charAt(index))/multiplier;
                multiplier *= 10;
                index++;
            }
        }
        return finalNumber * 1.0f;
    }

    private float CharToFloat(char charAt) {
        switch(charAt)
        {
            case '0':
                return 0.0f;
            case '1':
                return 1.0f;
            case '2':
                return 2.0f;
            case '3':
                return 3.0f;
            case '4':
                return 4.0f;
            case '5':
                return 5.0f;
            case '6':
                return 6.0f;
            case '7':
                return 7.0f;
            case '8':
                return 8.0f;
            case '9':
                return 9.0f;
            default:
                return -1.0f;
        }
    }

    private boolean isFloat(String tempFloatValue) {
        // If it contains a . or an e or neither but has 'f' at the end, it is considered a float
        for(int i = 0; i < tempFloatValue.length();i++)
        {
            if(tempFloatValue.charAt(i) == '.' || tempFloatValue.charAt(i) == 'e' || tempFloatValue.charAt(i) == 'E')
                return true;
        }
        char lastCharacter = tempFloatValue.charAt(tempFloatValue.length()-1);
        if(lastCharacter == 'f' || lastCharacter == 'F')
            return true;

        return false;
    }
}
